ktlint/src/main/kotlin/com/pinterest/ktlint/internal/FileUtils.kt

```diff
@@ -1,37 +1,116 @@
 package com.pinterest.ktlint.internal
 
-import com.github.shyiko.klob.Glob
 import com.pinterest.ktlint.core.KtLint
 import com.pinterest.ktlint.core.LintError
 import com.pinterest.ktlint.core.RuleSet
 import java.io.File
+import java.nio.file.FileSystem
+import java.nio.file.FileVisitResult
+import java.nio.file.Files
 import java.nio.file.Path
 import java.nio.file.Paths
+import java.nio.file.SimpleFileVisitor
+import java.nio.file.attribute.BasicFileAttributes
 import kotlin.system.exitProcess
 
 internal val workDir: String = File(".").canonicalPath
+private val tildeRegex = Regex("^(!)?~")
 
-internal fun List<String>.fileSequence(): Sequence<File> {
-    val kotlinFiles: Sequence<Path> = if (isEmpty()) {
-        Glob.from("**/*.kt", "**/*.kts")
-            .iterate(
-                Paths.get(workDir),
-                Glob.IterationOption.SKIP_HIDDEN
-            )
-            .asSequence()
+internal fun FileSystem.fileSequence(
+    globs: List<String>,
+    rootDir: Path = Paths.get(".").toAbsolutePath().normalize()
+): Sequence<Path> {
+    checkGlobsContainAbsolutePath(globs)
+
+    val pathMatchers = if (globs.isEmpty()) {
+        setOf(
+            getPathMatcher("glob:**$globSeparator*.kt"),
+            getPathMatcher("glob:**$globSeparator*.kts")
+        )
     } else {
-        // Converting List<String> to Array<String> and passing it to Glob.from(patterns) skips some files
-        // See https://github.com/pinterest/ktlint/issues/942
-        map {
-            Glob.from(expandTilde(it))
-                .iterate(Paths.get(workDir))
-        }
-            .asSequence()
-            .flatMap { it.asSequence() }
+        globs
+            .filterNot { it.startsWith("!") }
+            .map {
+                getPathMatcher(it.toGlob(rootDir))
+            }
+    }
+
+    val negatedPathMatchers = if (globs.isEmpty()) {
+        emptySet()
+    } else {
+        globs
+            .filter { it.startsWith("!") }
+            .map { getPathMatcher(it.removePrefix("!").toGlob(rootDir)) }
     }
 
-    return kotlinFiles
-        .map(Path::toFile)
+    val result = mutableListOf<Path>()
+    Files.walkFileTree(
```

Comment:
walking an entire huge project from the root to see what matches the globs is hugely inefficient.
Each glob should be expanded instead via Files.newDirectoryStream()